logbook-api/src/test/java/org/zalando/logbook/attributes/HttpAttributesTest.java

```diff
@@ -0,0 +1,75 @@
+package org.zalando.logbook.attributes;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+final class HttpAttributesTest {
+
+    private final Map<String, Object> mapWithTwoKeys = new HashMap<>();
+    private final Map<String, Object> mapWithOneKey = new HashMap<>();
+
+    {
+        mapWithTwoKeys.put("key1", "val1");
+        mapWithTwoKeys.put("key2", "val2");
+
+        mapWithOneKey.put("key", "val");
+    }
+
+    @Test
+    void emptyHttpAttributesShouldBeImmutable() {
+        final HttpAttributes attributes = HttpAttributes.EMPTY;
+
+        assertThat(attributes).isNotEqualTo(null);
+        assertThat(attributes).isEqualTo(Collections.emptyMap());
+        assertThat(Collections.emptyMap()).isEqualTo(attributes);
+
+        assertThat(attributes.isEmpty()).isTrue();
+        assertThat(attributes).isEqualTo(new HttpAttributes());
+        assertThat(attributes.hashCode()).isEqualTo(0);
+        assertThat(attributes.toString()).isEqualTo("{}");
+
+        //noinspection DataFlowIssue
+        assertThatThrownBy(() -> attributes.put("key", "val"))
+                .isInstanceOf(UnsupportedOperationException.class);
+    }
+
+    @Test
+    void singletonHttpAttributesShouldBeImmutable() {
+        final Map<String, Object> map1Clone = new HashMap<>(mapWithOneKey);
```

Comment:
Why do you make clones everywhere?

Reply:
Just wanted to make sure that the equality is based on the `map.equals(o)` and not object reference comparison `this == o`:

https://github.com/zalando/logbook/blob/b9a093dac25f6ddde152c00d018d99cb33218f4d/logbook-api/src/main/java/org/zalando/logbook/attributes/HttpAttributes.java#L30-L35

Reply:
I've used https://jqno.nl/equalsverifier/ before to check the full equals/hashCode contract. Maybe it's applicable here as well.

Reply:
Addressed in 22d7160. I still needed clones to assert immutability in some cases, but testing eaulas and hashCode is much better.